### ISSUE TITLE
Getting hidden states for specific tokens?

### ISSUE THREAD

**Issue by joshmyersdean** [allenai/molmo#16]
Hello!

Thank you for releasing this work. I was wondering if you had an easy way align hidden states with specific output tokens (e.g., `"="`)? Ideally we would do this without having to fully perform generation/decoding. Below is what I have tried so far, but the output is gibberish.

```python
# process the image and text
inputs = processor.process(
    images=[Image.open(requests.get("https://picsum.photos/id/237/536/354", stream=True).raw)],
    text="Point to each part of the dog. For each part, describe what it is."
)

# move inputs to the correct device and make a batch of size 1
inputs = {k: v.to(model.device).unsqueeze(0) for k, v in inputs.items()}

out = model.model(inputs['input_ids']) # model is Molmo 7B from HF

# below is taken from Molmo Trainer class
logits_for_loss = out.logits[..., :-1, :].contiguous()
# shape: (batch_size * seq_len, vocab_size)
logits_for_loss = logits_for_loss.view(-1, logits_for_loss.size(-1))

tokens = torch.argmax(logits_for_loss, dim=-1)

print(processor.tokenizer.decode(tokens)) 
>>>> '.NULL \n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n ( (,\n\n\n\n\n\n    ,\n ( ( (... ( ( ( (\n\n )\n ( ( ( ( ( ( ( (\n\n\n ? ( ( ( ( ( ( ( ( (\n\n alt ( ( ( ( ( ( ( ( ( ( ()\n photograph ( ( ( ( ( ( ( ( ( ()\n)\n, ( ( ( ( ( ( ( ( ( ( ( ( photograph ( ( ( ( ( ( ( ( ( ( ( ( photograph ( ( ( ( ( ( ( ( ( ( ( ()\n\n ( ( ( ( ( ( ( ( ( ( ( ()\n\n ( ( ( ( ( ( ( ( ( ( ( ()\n\n Thisy%="="="="="="="="="="="="="=")\n="="="=" "="="="="="="="="="="="="=")\n ( (="99)\n1="=" =="="=" ="="="=" ( ( (; ( (=" yellow)\n)\n=":="="=" (   ( ( ( ( ( ( ( ( ( bright:\n:\n)\n=": (=" bright ( ( ( ( ( ( ( ( ( ( ( ( (:\n:\n)\n ( ( ( bright bright ( ( ( ( ( ( ( ( ( ( ( ( (:\n)\n)\n ( ( ( \' bright ( ( ( ( ( ( ( ( ( ( ( ( ( bright)\n)\n ( ( ( ( bright ( ( ( ( ( ( ( ( ( ( ( ( (:\n)\n... ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( bright (... ( ( ( ( outdoor ( ( ( ( ( ( ( ( ( ( ( ( (:\n)\n... ( ( ( ( bright ( ( ( ( ( ( ( ( ( ( ( ( ()\n)\n. ( ( ( ( bright ( ( ( ( ( ( ( ( ( ( ( ( ( ()\n. ( ( ( ( outdoor ( ( ( ( ( ( ( ( ( ( ( ( ()\n)\n)\n\n ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( ()\n)\n)\n\n ( ( ( ( bright ( ( ( ( ( ( ( ( ( ( ( ( ( ()\n; ( ( ( bright outdoor ( ( ( ( ( ( ( ( ( ( ( ( (\n (; ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( ( (;\n ( ( ( bright ( ( ( ( ( ( ( ( ( ( ( ( ( ( (; ( ( ( ( outdoor ( ( ( ( ( ( ( ( ( ( ( ( (\n\n)\n ( ( ( ( bright ( ( ( ( ( ( ( ( ( ( ( ( ( ( ()\n This: What x flower flower of the roof\'s There how part, list it it is. Count <'=
```

where generation gives

```python
output = model.generate_from_batch(
    inputs,
    GenerationConfig(max_new_tokens=200, stop_strings="<|endoftext|>"),
    tokenizer=processor.tokenizer
)

# only get generated tokens; decode them to text
generated_tokens = output[0,inputs['input_ids'].size(1):]
generated_text = processor.tokenizer.decode(generated_tokens, skip_special_tokens=True)

print(generated_text)
>>>> <points x1="26.0" y1="36.0" x2="38.6" y2="31.5" x3="46.5" y3="85.0" x4="51.0" y4="42.0" x5="57.0" y5="26.5" x6="59.0" y6="62.5" x7="60.5" y7="78.5" x8="61.0" y8="13.0" x9="65.5" y9="40.0" x10="73.5" y10="26.5" alt="part of the dog. For each part, describe what it is.">part of the dog. For each part, describe what it is.</points>
```

My idea is that once I know the decoding outputs align, I can just create a mask for the hidden state based off the token indices I care about.

Thank you for your time!

**Comment by joshmyersdean** [comment]
This was a fundamental misunderstanding on my part!
